src/nnet3/nnet-simple-component.cc

```diff
@@ -668,12 +718,99 @@ void ClipGradientComponent::Backprop(const std::string &debug_info,
       in_deriv->ApplyCeiling(clipping_threshold_);
       in_deriv->ApplyFloor(-1 * clipping_threshold_);
     }
+
+    to_update->num_backproped_ += 1;
+    RepairGradients(debug_info, in_value, in_deriv, to_update);
   }
 }
 
+void ClipGradientComponent::RepairGradients(
+    const std::string &debug_info,
+    const CuMatrixBase<BaseFloat> &in_value,
+    CuMatrixBase<BaseFloat> *in_deriv, ClipGradientComponent *to_update) const {
+  KALDI_ASSERT(to_update != NULL);
+  BaseFloat default_clipped_proportion_threshold = 0.01;
+  BaseFloat default_clipping_to_value = 0.0;
+
+  // we use this 'repair_probability' (hardcoded for now) to limit
+  // this code to running on about half of the minibatches.
+  BaseFloat repair_probability = 0.5;
+  if (self_repair_clipped_proportion_threshold_ >= 1.0 ||
+      self_repair_scale_ == 0.0 || count_ == 0 ||
+      RandUniform() > repair_probability)
+    return;
+
+  BaseFloat unset = kUnsetValue; // -1000.0
+
+  BaseFloat clipped_proportion_threshold =
+      (self_repair_clipped_proportion_threshold_ == unset ?
+       default_clipped_proportion_threshold :
+       self_repair_clipped_proportion_threshold_);
+  BaseFloat clipping_to_value =
+      (self_repair_clipping_to_value_ == unset ?
+       default_clipping_to_value :
+       self_repair_clipping_to_value_);
+  KALDI_ASSERT(clipping_to_value >= 0.0 && self_repair_scale_ > 0.0);
+
+  BaseFloat clipped_proportion =
+    (count_ > 0 ? static_cast<BaseFloat>(num_clipped_) / count_ : 0);
+  // in-deriv would be modified only when clipped_proportion exceeds the
+  // threshold
+  if (clipped_proportion <= clipped_proportion_threshold)
+    return;
+
+  to_update->num_self_repaired_ += 1;
+  if (to_update->debug_info_ == "") //debug
+    to_update->debug_info_ = debug_info; //debug
+  if (to_update->num_self_repaired_ == 1)
+    KALDI_LOG << "ClipGradientComponent(node_name=" << debug_info
+              << ")'s self-repair was activated as the first time at the "
+              << num_backproped_
+              << "-th call of Backprop() in this training job.";
+
+  // sign_mat = sign(in_value), i.e.,
+  // An element in sign_mat is 1 if its corresponding element in in_value > 0,
+  // or -1 otherwise
+  CuMatrix<BaseFloat> sign_mat(in_value);
+  sign_mat.ApplyHeaviside();
+  sign_mat.Scale(2.0);
+  sign_mat.Add(-1.0);
+
+  // repair_mat =
+  // floor(abs(in_value) - clipping_to_value, 0) .* sign(in_value)
+  CuMatrix<BaseFloat> repair_mat(in_value);
+  repair_mat.ApplyPowAbs(1.0);
+  repair_mat.Add(-clipping_to_value);
+  repair_mat.ApplyFloor(0.0);
+  repair_mat.MulElements(sign_mat);
+
+  // magnitude =
```

Comment:
I think it would be better to set the magnitude at the whole minibatch level (or to be more precise, at the whole matrix level) because that will tend to give higher magnitudes of self-repair to input vectors that have larger absolute value, which will tend to be the ones that are diverging.